### ISSUE TITLE
Upgrader - "Full" fixed-length fields can abort UTF8 conversion

### ISSUE THREAD

**Issue by sbulen** [SimpleMachines/SMF#4019]
In instances where we have a fixed text length field, the UTF8 conversion can fail if the text field is full.  

I am having a problem with the "reason" field on ban_groups, which is varchar(255).  Someone copied the offending message in the "reason" field, and it truncated at 255 characters.  The text includes some non-ascii characters, e.g., umlauts, and some special characters, e.g., quotes.  The upgrade proceeds OK if I edit this particular column on this particular record first.  

The UTF8 conversion intends to do the conversion in 2 steps: first to a binary field type, then to the target field type.  When attempting to convert the varchar(255) to varbinary(255) the UTF8 conversion fails on this record, and aborts the whole conversion & upgrade.  

The code that performs this is in the ~2793 - 2820 range:
https://github.com/SimpleMachines/SMF2.1/blob/release-2.1/other/upgrade.php#L2793
```
			// Only change the column if the data doesn't match the current charset.
			if ((count($table_charsets) === 1 && key($table_charsets) !== $charsets[$upcontext['charset_detected']]) || count($table_charsets) > 1)
			{
				$updates_blob = '';
				$updates_text = '';
				foreach ($table_charsets as $charset => $columns)
				{
					if ($charset !== $charsets[$upcontext['charset_detected']])
					{
						foreach ($columns as $column)
						{
							$updates_blob .= '
								CHANGE COLUMN `' . $column['Field'] . '` `' . $column['Field'] . '` ' . strtr($column['Type'], array('text' => 'blob', 'char' => 'binary')) . ($column['Null'] === 'YES' ? ' NULL' : ' NOT NULL') . (strpos($column['Type'], 'char') === false ? '' : ' default \'' . $column['Default'] . '\'') . ',';
							$updates_text .= '
								CHANGE COLUMN `' . $column['Field'] . '` `' . $column['Field'] . '` ' . $column['Type'] . ' CHARACTER SET ' . $charsets[$upcontext['charset_detected']] . ($column['Null'] === 'YES' ? '' : ' NOT NULL') . (strpos($column['Type'], 'char') === false ? '' : ' default \'' . $column['Default'] . '\'') . ',';
						}
					}
				}
				// Change the columns to binary form.
				$smcFunc['db_query']('', '
					ALTER TABLE {raw:table_name}{raw:updates_blob}',
					array(
						'table_name' => $table_info['Name'],
						'updates_blob' => substr($updates_blob, 0, -1),
					)
				);

```

**Comment by sbulen** [comment]
My understanding of what it is trying to do, based on some reading...  A simple latin1 => utf8 conversion works great with an ALTER TABLE CHANGE COLUMN - when the source data is all latin1.  The problem is most latin1 DBs have non-ascii characters in them.  The conversion can actually corrupt well-formed utf8 already in a latin1 table.  To protect against this, it is recommended to convert first to binary, because the binary => utf8 conversion works perfectly, no matter what is in the source.  So these are usually done in two steps - bouncing off of binary.  This explains why there is a binary conversion in there...

An excellent article that explains the problem & solutions for noobs, including sample code:
http://nicj.net/mysql-converting-an-incorrect-latin1-column-to-utf8/
https://github.com/nicjansma/mysql-convert-latin1-to-utf8

HOWEVER...   I do have issues with this implementation in SMF...  It is attempting to convert everything to "charset_detected", not UTF8.  charset_detected is usually latin1 on a DB that needs upgrading.  So where it should be going from latin1=>binary, then binary =>utf8, instead, it goes from latin1=>binary then binary =>latin1...  A _subsequent_ step actually changes the setting (to UTF8, hardcoded, lines ~2850+) at the table level, and this appears to work fine.  No data corruption.  I have even tested it with UTF8 content in a latin1 table.  It's working, so I have left it alone...

**Comment by sbulen** [comment]
You get a white screen with a single error message: "Data too long for column 'reason' at row 1".

The contents of the offending 'reason' in ban_groups:
"SPAM-copy of his post.&quot;In Australien gibt es einige Handyanbieter. Je nachdem wo du hin willst kommt es drauf an welchen Anbieter du dir aussuchst. Falls du planst viel in der Pampa rumzureisen, würde ich dir auf jeden Fall ein Telstra Handy empfehle"

**Comment by sbulen** [comment]
I have tested a variety of scenarios, and I can reproduce this issue at will on my "manually" converted UTF8 database (utf8 tables, utf8 content, SMF global_character_set setting not utf8).  This DB emulates a forum where the site admin performed the UTF8 conversion without using the SMF tool.  

I can NOT reproduce this at will in a latin 1 database (latin1 tables, utf8 content,  SMF global_character_set setting not utf8).

So...  A proper fix would be to NOT re-encode UTF8 columns that are already UTF8.   I already have a PR out there for that:   Do not re-encode data already utf8 #3984

I have tested & confirmed that this issue can be closed when #3984 goes in.
